"ERROR Resource path must not end with '/' but is '.../'" must be improved

Currently if one referenced resource ends with a slash the following exception is logged
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  49.602 s
[INFO] Finished at: 2020-03-12T09:33:12+01:00
[INFO] ------------------------------------------------------------------------
[DEBUG] No threads registered for execution at the end of the build
[ERROR] Failed to execute goal org.apache.jackrabbit:filevault-package-maven-plugin:1.1.0:validate-package (default-validate-package) on project nc-website-components-package: Execution default-validate-package of goal org.apache.jackrabbit:filevault-package-maven-plugin:1.1.0:validate-package failed: Resource path must not end with '/' but is '..../' -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.jackrabbit:filevault-package-maven-plugin:1.1.0:validate-package (default-validate-package) on project nc-website-components-package: Execution default-validate-package of goal org.apache.jackrabbit:filevault-package-maven-plugin:1.1.0:validate-package failed: Resource path must not end with '/' but is '/libs/core/wcm/components/teaser/v1/teaser/
```

This error message must be improved to at least emit the file path where the error occurred!
Also the debug log level does not help much as the previous output is 
`[DEBUG] ValidationViolation: "Validating jcr file '<some jcr file>'..."` which is not the failing file but the previous one!

The issue can be easily reproduced by setting the value `sling:resourceSuperType` on a component definition to a value like `core/wcm/components/teaser/v1/teaser/`

The situation will improve with the upcoming FileVault 3.4.4 (https://issues.apache.org/jira/browse/JCRVLT-406)

--- a/aem-classification-validator/src/main/java/biz/netcentric/filevault/validator/aem/classification/AemClassificationValidator.java
+++ b/aem-classification-validator/src/main/java/biz/netcentric/filevault/validator/aem/classification/AemClassificationValidator.java
@@ -41,6 +41,7 @@ import org.apache.jackrabbit.vault.validation.spi.ValidationMessageSeverity;
 import org.apache.sling.api.SlingConstants;
 import org.apache.sling.jcr.resource.api.JcrResourceConstants;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 public class AemClassificationValidator implements DocumentViewXmlValidator, GenericJcrDataValidator, NodePathValidator {
 
@@ -77,7 +78,13 @@ public class AemClassificationValidator implements DocumentViewXmlValidator, Gen
     private static final String SLING_RESOURCE_TYPE_PROPERTY_NAME = NameFactoryImpl.getInstance().create(JcrResourceConstants.SLING_NAMESPACE_URI,SlingConstants.PROPERTY_RESOURCE_TYPE).toString();
     private static final String SLING_RESOURCE_SUPER_TYPE_PROPERTY_NAME = NameFactoryImpl.getInstance().create(JcrResourceConstants.SLING_NAMESPACE_URI,SlingConstants.PROPERTY_RESOURCE_SUPER_TYPE).toString();
     
-    
+
+    public static final String MESSAGE_SUBJECT_NODE = "Element with name \"%s\"";
+    public static final String MESSAGE_SUBJECT_FILE = "This file";
+
+    /** Pattern to be used with {@link String#format(String, Object...)} */
+    static final String VIOLATION_MESSAGE_STRING = "%s %s resource '%s' which is marked as '%s'. It therefore violates the content classification!";
+
     private final ContentClassificationMapper classificationMap;
     private final Collection<String> whitelistedResourcePaths;
     private final Collection<Pattern> whitelistedResourcePathPatterns;
@@ -103,14 +110,14 @@ public class AemClassificationValidator implements DocumentViewXmlValidator, Gen
 
     @Override
     public Collection<ValidationMessage> validate(@NotNull String path) {
-        // check overlay usage in addition for non-docview files
-        OverlayViolation violation = getOverlayViolation(path);
-        if (violation != null && !overlaidNodePaths.contains(path)) {
-            overlaidNodePaths.add(path);
-            return Collections.singleton(getSimpleFileViolationMessage(ContentUsage.OVERLAY, violation.overlaidResourceType, violation.classification, violation.remark));
-        } else {
-            return null;
+        if (!overlaidNodePaths.contains(path)) {
+            // check overlay usage in addition for non-docview files
+            ValidationMessage message = validateClassification(path,  ContentUsage.OVERLAY, MESSAGE_SUBJECT_FILE);
+            if (message != null) {
+               return Collections.singleton(message);
+            }
         }
+        return null;
     }
 
     @Override
@@ -138,7 +145,7 @@ public class AemClassificationValidator implements DocumentViewXmlValidator, Gen
                 Matcher matcher = regex.matcher(line);
                 while (matcher.find()) {
                     String resourceType = matcher.group(1);
-                    ValidationMessage message = validateResourceTypeUsage(resourceType);
+                    ValidationMessage message = validateClassification(resourceType, ContentUsage.REFERENCE, MESSAGE_SUBJECT_FILE);
                     if (message != null) {
                         messages.add(message);
                     }
@@ -151,92 +158,64 @@ public class AemClassificationValidator implements DocumentViewXmlValidator, Gen
     @Override
     public Collection<ValidationMessage> validate(@NotNull DocViewNode node, @NotNull String nodePath, @NotNull Path filePath, boolean isRoot) {
         Collection<ValidationMessage> messages = new LinkedList<>();
+        String subject = String.format(MESSAGE_SUBJECT_NODE, node.label);
+
         // attributes resourceType ...
         String usedResource = node.getValue(SLING_RESOURCE_TYPE_PROPERTY_NAME);
-        if (usedResource != null) {
-            Entry<ContentClassification,String> classificationAndRemark = classificationMap.getContentClassificationAndRemarkForResourcePath(usedResource, whitelistedResourcePathPatterns);
-            if (!classificationAndRemark.getKey().isAllowed(ContentUsage.REFERENCE)) {
-                messages.add(getDocviewViolationMessage(node.label, ContentUsage.REFERENCE, usedResource, classificationAndRemark.getKey(), classificationAndRemark.getValue()));
-            }
+        ValidationMessage message = validateClassification(usedResource,  ContentUsage.REFERENCE, subject);
+        if (message != null) {
+           messages.add(message);
         }
 
         // ... and resourceSuperType are considered
         String superResource = node.getValue(SLING_RESOURCE_SUPER_TYPE_PROPERTY_NAME);
-        if (superResource != null) {
-            Entry<ContentClassification,String> classificationAndRemark = classificationMap.getContentClassificationAndRemarkForResourcePath(superResource, whitelistedResourcePathPatterns);
-            if (!classificationAndRemark.getKey().isAllowed(ContentUsage.INHERIT)) {
-                messages.add(getDocviewViolationMessage(node.label, ContentUsage.INHERIT, superResource, classificationAndRemark.getKey(), classificationAndRemark.getValue()));
-            }
+        message = validateClassification(superResource,  ContentUsage.INHERIT, subject);
+        if (message != null) {
+           messages.add(message);
         }
 
         // check overlays in addition
-        OverlayViolation violation = getOverlayViolation(nodePath);
-        if (violation != null && !overlaidNodePaths.contains(nodePath)) {
-            messages.add(getDocviewViolationMessage(node.label, ContentUsage.OVERLAY, violation.overlaidResourceType, violation.classification, violation.remark));
-            overlaidNodePaths.add(nodePath);
+        message = validateClassification(nodePath,  ContentUsage.OVERLAY, subject);
+        if (message != null) {
+           messages.add(message);
+           overlaidNodePaths.add(nodePath);
         }
+        
         // TODO: check usage of clientlib dependencies/embeds
         return messages;
     }
 
-    private final class OverlayViolation {
-        private final ContentClassification classification;
-        private final String overlaidResourceType;
-        private final String remark;
-        
-        public OverlayViolation(ContentClassification classification, String overlaidResourceType, String remark) {
-            super();
-            this.classification = classification;
-            this.overlaidResourceType = overlaidResourceType;
-            this.remark = remark;
-        }
-    }
-    /**
-     * 
-     * @param path
-     * @return the overlaid path and a classification for that path only in case the given overlay violates it, otherwise {@code null}.
-     */
-    private OverlayViolation getOverlayViolation(String path) {
-        if (!path.startsWith("/apps/")) {
-            return null; // this is not an overlay at all, therefore no violation
-        }
-        // is this an overlay?
-        String overlaidResource = "/libs/" + path.substring("/apps/".length());
-        Entry<ContentClassification, String> classificationAndRemark = classificationMap.getContentClassificationAndRemarkForResourcePath(overlaidResource, whitelistedResourcePathPatterns);
-        if (classificationAndRemark.getKey().isAllowed(ContentUsage.OVERLAY)) {
+    private @Nullable ValidationMessage validateClassification(String resourcePath, ContentUsage usage, String messageSubject) {
+        if (resourcePath == null) {
+            // skip validation
             return null;
         }
-        return new OverlayViolation(classificationAndRemark.getKey(), overlaidResource, classificationAndRemark.getValue());
+        if (resourcePath.endsWith("/") && !resourcePath.equals("/")) {
+            // add subject and usage to message
+            return new ValidationMessage(defaultSeverity, "Resource path must not end with '/' but is '" + resourcePath + "'");
+        }
+        
+        if (usage == ContentUsage.OVERLAY) {
+            if (!resourcePath.startsWith("/apps/")) {
+                return null; // this is not an overlay at all, therefore no violation
+            }
+            // is this an overlay?
+            resourcePath = "/libs/" + resourcePath.substring("/apps/".length());
+        }
+        Entry<ContentClassification,String> classificationAndRemark = classificationMap.getContentClassificationAndRemarkForResourcePath(resourcePath, whitelistedResourcePathPatterns);
+        ContentClassification classification = classificationAndRemark.getKey();
+        if (!classificationAndRemark.getKey().isAllowed(usage)) {
+            return new ValidationMessage(getSeverityForClassification(classification), extendMessageWithRemark(String.format(VIOLATION_MESSAGE_STRING, messageSubject, usage.getLabel(), resourcePath, classification.getLabel()), classificationAndRemark.getValue()));
+        }
+        return null;
     }
 
-    private ValidationMessage validateResourceTypeUsage(String resourceType) {
-        Entry<ContentClassification,String> classificationAndRemark = classificationMap.getContentClassificationAndRemarkForResourcePath(resourceType, whitelistedResourcePathPatterns);
-        if (!classificationAndRemark.getKey().isAllowed(ContentUsage.REFERENCE)) {
-            return getSimpleFileViolationMessage(ContentUsage.REFERENCE, resourceType, classificationAndRemark.getKey(), classificationAndRemark.getValue());
-        } else {
-            return null;
-        }
-    }
-    
     private static boolean isHtlFile(Path file) {
         return HTL_PATH_MATCHER.matches(file);
     }
 
     private static boolean isJspFile(Path file) {
         return JSP_PATH_MATCHER.matches(file);
-    }
-
-    /** Pattern to be used with {@link String#format(String, Object...)} */
-    static final String DOCVIEW_VIOLATION_MESSAGE_STRING = "Element with name \"%s\" %s resource '%s' which is marked as '%s'. It therefore violates the content classification!";
-    static final String SIMPLEFILE_VIOLATION_MESSAGE_STRING = "This file %s resource '%s' which is marked as '%s'. It therefore violates the content classification!";
-    @NotNull ValidationMessage getDocviewViolationMessage(String label, ContentUsage usage, String targetResourceType, ContentClassification classification, String remark) {
-        String message = extendMessageWithRemark(String.format(DOCVIEW_VIOLATION_MESSAGE_STRING, label, usage.getLabel(), targetResourceType, classification.getLabel()), remark);
-        return new ValidationMessage(getSeverityForClassification(classification), message);
-    }
-
-    
-    @NotNull ValidationMessage getSimpleFileViolationMessage(ContentUsage usage, String resourceType, ContentClassification classification, String remark) {
-        return new ValidationMessage(getSeverityForClassification(classification), extendMessageWithRemark(String.format(SIMPLEFILE_VIOLATION_MESSAGE_STRING, usage.getLabel(), resourceType, classification.getLabel()), remark));
     }
 
     static @NotNull String extendMessageWithRemark(@NotNull String message, String remark) {

--- a/aem-classification-validator/src/main/java/biz/netcentric/filevault/validator/aem/classification/ContentClassificationMapper.java
+++ b/aem-classification-validator/src/main/java/biz/netcentric/filevault/validator/aem/classification/ContentClassificationMapper.java
@@ -29,7 +29,7 @@ public interface ContentClassificationMapper {
 
     /**
      * Returns the classification for the given resource type.
-     * In case the given {@code resourcePath} matches any of the {@code whiteLilstedResourcePaths} it returns PUBLIC.
+     * In case the given {@code resourcePath} matches any of the {@code whitelistedResourcePaths} it returns PUBLIC.
      * @param resourcePath the absolute resource path
      * @param whitelistedResourcePaths the whitelisted resource paths as regular expression patterns. 
      * @return the classification and the optional remark belonging to the given resource type

--- a/aem-classification-validator/src/test/java/biz/netcentric/filevault/validator/aem/classification/AemClassificationValidatorTest.java
+++ b/aem-classification-validator/src/test/java/biz/netcentric/filevault/validator/aem/classification/AemClassificationValidatorTest.java
@@ -42,6 +42,8 @@ import org.apache.jackrabbit.vault.validation.spi.ValidationMessageSeverity;
 import org.apache.jackrabbit.vault.validation.spi.ValidatorSettings;
 import org.apache.jackrabbit.vault.validation.spi.impl.DocumentViewParserValidator;
 import org.apache.jackrabbit.vault.validation.spi.impl.DocumentViewParserValidatorFactory;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.xml.sax.SAXException;
@@ -118,7 +120,14 @@ public class AemClassificationValidatorTest {
          );
     }
 
-  
+    @Test
+    public void testReferencingWithTrailingSlashesInDocviewXml()
+            throws SAXException, IOException, ParserConfigurationException, URISyntaxException, FileSystemException {
+        Assertions.assertEquals(Collections.singletonList(
+                new ValidationViolation("myId", ValidationMessageSeverity.ERROR, "Resource path must not end with '/' but is 'core/wcm/components/teaser/v1/teaser/'", EXAMPLE_DOCVIEW_PATH, Paths.get(""), "/apps/example", 8, 107, null)
+                ), validateJcrDocView(validator, "myId", "/invalid-resource-type.xml", EXAMPLE_DOCVIEW_PATH));
+    }
+
     @Test
     public void testInheritingViolationsInDocviewXml()
             throws SAXException, IOException, ParserConfigurationException, URISyntaxException, FileSystemException {
@@ -156,10 +165,10 @@ public class AemClassificationValidatorTest {
         assertNull(validator.validate("/apps/abstract"));
         assertNull(validator.validate("/apps/abstract/test11"));
         assertNull(validator.validate("/apps/abstract/test12"));
-        assertEquals(Collections.singleton(validator.getSimpleFileViolationMessage(ContentUsage.OVERLAY, "/libs/final", ContentClassification.FINAL, "finalremark")), validator.validate("/apps/final"));
-        assertEquals(Collections.singleton(validator.getSimpleFileViolationMessage(ContentUsage.OVERLAY, "/libs/final/test21", ContentClassification.INTERNAL_CHILD, "finalremark")), validator.validate("/apps/final/test21"));
-        assertEquals(Collections.singleton(validator.getSimpleFileViolationMessage(ContentUsage.OVERLAY, "/libs/internal", ContentClassification.INTERNAL, "internalremark")), validator.validate("/apps/internal"));
-        assertEquals(Collections.singleton(validator.getSimpleFileViolationMessage(ContentUsage.OVERLAY, "/libs/internal/test21",ContentClassification.INTERNAL,  "internalremark")), validator.validate("/apps/internal/test21"));
+        assertEquals(Collections.singleton(getSimpleFileViolationMessage(ValidationMessageSeverity.ERROR, ContentUsage.OVERLAY, "/libs/final", ContentClassification.FINAL, "finalremark")), validator.validate("/apps/final"));
+        assertEquals(Collections.singleton(getSimpleFileViolationMessage(ValidationMessageSeverity.ERROR, ContentUsage.OVERLAY, "/libs/final/test21", ContentClassification.INTERNAL_CHILD, "finalremark")), validator.validate("/apps/final/test21"));
+        assertEquals(Collections.singleton(getSimpleFileViolationMessage(ValidationMessageSeverity.ERROR, ContentUsage.OVERLAY, "/libs/internal", ContentClassification.INTERNAL, "internalremark")), validator.validate("/apps/internal"));
+        assertEquals(Collections.singleton(getSimpleFileViolationMessage(ValidationMessageSeverity.ERROR, ContentUsage.OVERLAY, "/libs/internal/test21",ContentClassification.INTERNAL,  "internalremark")), validator.validate("/apps/internal/test21"));
         assertNull(validator.validate("/apps/public"));
         assertNull(validator.validate("/apps/public/test41"));
 
@@ -172,7 +181,7 @@ public class AemClassificationValidatorTest {
         try (InputStream input = this.getClass().getClassLoader().getResourceAsStream("htl-example.html")) {
             Collection<ValidationMessage> messages = validator.validateJcrData(input, SIMPLEFILE_HTL_PATH, new HashMap<String, Integer>());
             // and check violations
-            assertEquals(Collections.singletonList(validator.getSimpleFileViolationMessage(ContentUsage.REFERENCE, "/libs/abstract/test",  ContentClassification.ABSTRACT, "abstractremark")), messages);
+            assertEquals(Collections.singletonList(getSimpleFileViolationMessage(ValidationMessageSeverity.ERROR, ContentUsage.REFERENCE, "/libs/abstract/test",  ContentClassification.ABSTRACT, "abstractremark")), messages);
         }
     }
 
@@ -204,7 +213,7 @@ public class AemClassificationValidatorTest {
         Collection<ValidationMessage> actualMessages = validateJcrDocView(validator, "myid", name, path);
         Collection<ValidationViolation> expectedMessages = Arrays.stream(violations).map(a -> 
             ValidationViolation.wrapMessage("myid", 
-                    validator.getDocviewViolationMessage(a.name, a.usage, a.targetResourceType, a.classification, a.remark),
+                    getDocviewViolationMessage(ValidationMessageSeverity.ERROR, a.name, a.usage, a.targetResourceType, a.classification, a.remark),
                     path, Paths.get(""), a.nodePath, a.line, a.column)).collect(Collectors.toList());
         
         assertEquals(expectedMessages, actualMessages);
@@ -219,5 +228,18 @@ public class AemClassificationValidatorTest {
             Collection<ValidationMessage> allMessages = docViewValidator.validateJcrData(input, filePath, nodePathsAndLineNumbers);
             return allMessages.stream().filter(a -> a.getSeverity().ordinal() >= ValidationMessageSeverity.ERROR.ordinal()).collect(Collectors.toList());
         }
+    }
+
+    /** Pattern to be used with {@link String#format(String, Object...)} */
+    static final String DOCVIEW_VIOLATION_MESSAGE_STRING = "Element with name \"%s\" %s resource '%s' which is marked as '%s'. It therefore violates the content classification!";
+    static final String SIMPLEFILE_VIOLATION_MESSAGE_STRING = "This file %s resource '%s' which is marked as '%s'. It therefore violates the content classification!";
+    static @NotNull ValidationMessage getDocviewViolationMessage(ValidationMessageSeverity severity, String label, ContentUsage usage, String targetResourceType, ContentClassification classification, String remark) {
+        String message = AemClassificationValidator.extendMessageWithRemark(String.format(DOCVIEW_VIOLATION_MESSAGE_STRING, label, usage.getLabel(), targetResourceType, classification.getLabel()), remark);
+        return new ValidationMessage(severity, message);
+    }
+
+    
+    static @NotNull ValidationMessage  getSimpleFileViolationMessage(ValidationMessageSeverity severity, ContentUsage usage, String resourceType, ContentClassification classification, String remark) {
+        return new ValidationMessage(severity, AemClassificationValidator.extendMessageWithRemark(String.format(SIMPLEFILE_VIOLATION_MESSAGE_STRING, usage.getLabel(), resourceType, classification.getLabel()), remark));
     }
 }

--- a/aem-classification-validator/src/test/resources/invalid-resource-type.xml
+++ b/aem-classification-validator/src/test/resources/invalid-resource-type.xml
@@ -1,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    jcr:primaryType="cq:Component"
+    componentGroup="heroparsys"
+    jcr:title="NC Hero"
+    imageDelegate="nc/website/components/content/image"
+    sling:resourceSuperType="core/wcm/components/teaser/v1/teaser/"
+    jcr:description="Displays a big headline, a subtitle and one or two buttons with a background image"/>

